flink-staging/flink-streaming/flink-streaming-core/src/main/java/org/apache/flink/streaming/api/streamvertex/StreamVertex.java

```diff
@@ -283,16 +288,18 @@ public StreamConfig getConfig() {
 	 * 
 	 * @param id
 	 */
-	private void actOnBarrier(long id) {
-		try {
-			outputHandler.broadcastBarrier(id);
-			// TODO checkpoint state here
-			confirmBarrier(id);
-			if (LOG.isDebugEnabled()) {
-				LOG.debug("Superstep " + id + " processed: " + StreamVertex.this);
+	private synchronized void actOnBarrier(long id) {
+		if (this.isRunning) {
+			try {
+				outputHandler.broadcastBarrier(id);
+				// TODO checkpoint state here
+				confirmBarrier(id);
+				if (LOG.isDebugEnabled()) {
+					LOG.debug("Superstep " + id + " processed: " + StreamVertex.this);
+				}
+			} catch (Exception e) {
+				// TODO:Figure this out properly
```

Comment:
I had some weird exceptions after the streams finished, and in any case this shouldnt affect the recovery logic